src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -0,0 +1,518 @@
+          {
+            float const candidate_dist = _metric(
+                leaf_permutation_i,
+                HappyTreeFriends::getLeafPermutationIndex(_bvh, left_child),
+                distance_left);
+            WeightedEdge const candidate_edge{i, left_child, candidate_dist};
+            if (candidate_edge < current_best)
+            {
+              current_best = candidate_edge;
+              Kokkos::atomic_min(&radius, candidate_dist);
+            }
+          }
+          else
+          {
+            traverse_left = true;
+          }
+        }
+
+        // Note: radius may have been already updated here from the left child
+        if (predicate(right_child) && distance_right <= radius)
+        {
+          if (HappyTreeFriends::isLeaf(_bvh, right_child))
+          {
+            float const candidate_dist = _metric(
+                leaf_permutation_i,
+                HappyTreeFriends::getLeafPermutationIndex(_bvh, right_child),
+                distance_right);
+            WeightedEdge const candidate_edge{i, right_child, candidate_dist};
+            if (candidate_edge < current_best)
+            {
+              current_best = candidate_edge;
+              Kokkos::atomic_min(&radius, candidate_dist);
+            }
+          }
+          else
+          {
+            traverse_right = true;
+          }
+        }
+      }
+
+      if (!traverse_left && !traverse_right)
+      {
+        node = *--stack_ptr;
+#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+        if (node != SENTINEL)
+        {
+          // This is a theoretically unnecessary duplication of distance
+          // calculation for stack nodes. However, for Cuda it's better than
+          // putting the distances in stack.
+          distance_node = distance(node);
+        }
+#else
+        distance_node = *--stack_distance_ptr;
+#endif
+      }
+      else
+      {
+        node = (traverse_left &&
+                (distance_left <= distance_right || !traverse_right))
+                   ? left_child
+                   : right_child;
+        distance_node = (node == left_child ? distance_left : distance_right);
+        if (traverse_left && traverse_right)
+        {
+          *stack_ptr++ = (node == left_child ? right_child : left_child);
+#if !defined(__CUDA_ARCH__) && !defined(__HIP_DEVICE_COMPILE__)
+          *stack_distance_ptr++ =
+              (node == left_child ? distance_right : distance_left);
+#endif
+        }
+      }
+    } while (node != SENTINEL);
+
+    // This check is only here to reduce hammering the atomics for large
+    // components. Otherwise, for a large number of points and a small number of
+    // components it becomes extremely expensive.
+    auto &component_edge = _edges(component - n + 1);
+    if (current_best < component_edge)
+    {
+      Kokkos::atomic_min(&component_edge, current_best);
+    }
+  }
+};
+
+template <class ExecutionSpace, class BVH, class Labels, class Edges,
```

Comment:
```suggestion
// For every component C, find the shortest edge (v, w) such that v is in C
// and w is not in C. The found edge is stored in component_out_edges(C).
template <class ExecutionSpace, class BVH, class Labels, class Edges,
```
Maybe `component_out_edges` should be the last argument, as it is the output, while everything else is input (except radius, which is kinda mutable).